### PR TITLE
Feat/issues/1096 kafka consumer

### LINKED ISSUE #1096: Spring Boot - Kafka - Dummy Consumer
Add a generic and a dummy consumer to generated code through a new endpoint.

@pascalgrimaud Can you assign me this one please?

### PR DESCRIPTION
**Pull request by fdelbrayelle**
fix #1096

### CODE REVIEW

**Comment by fdelbrayelle** [comment]
Remaining work...

```
[WARNING] Rule violated for class tech.jhipster.kafkaapp.dummy.infrastructure.primary.kafka.consumer.AbstractConsumer: lines covered ratio is 0.85, but expected minimum is 1.00
[WARNING] Rule violated for class tech.jhipster.kafkaapp.dummy.infrastructure.primary.kafka.consumer.AbstractConsumer: branches covered ratio is 0.66, but expected minimum is 1.00
```

Could you start reviewing the code please @pascalgrimaud?

File: src/main/resources/generator/server/springboot/broker/kafka/DummyConsumerIT.java.mustache [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import {{packageName}}.IntegrationTest;
+import {{packageName}}.dummy.infrastructure.secondary.kafka.producer.DummyProducer;
+
+@IntegrationTest
+class DummyConsumerIT {
+
+  @Autowired
+  private DummyProducer dummyProducer;
```
Comment (pascalgrimaud) [inline src/main/resources/generator/server/springboot/broker/kafka/DummyConsumerIT.java.mustache:15]: By using DummyProducer here, it means DummyConsumer can't be used without DummyProducer

2 solutions here:
- 1 single API to generate both: producer and consumer
- change the test here to not use DummyProducer

File: src/main/resources/generator/server/springboot/broker/kafka/DummyConsumerIT.java.mustache [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+  @Test
+  void shouldHandleMessage() {
+    final String messageToSend = "dummy";
+    dummyProducer.send(messageToSend);
+
+    ConsumerRecord<String, String> record = new ConsumerRecord<>("queue.kafkaapp.dummy", 0, 0, null, messageToSend);
+    boolean actualResult = dummyConsumer.handleMessage(record);
+
+    assertEquals(true, actualResult);
```
Comment (pascalgrimaud) [inline src/main/resources/generator/server/springboot/broker/kafka/DummyConsumerIT.java.mustache:28]: Use assertj

File: src/main/resources/generator/server/springboot/broker/kafka/DummyConsumerTest.java.mustache [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    Method postConstructDestroy = DummyConsumer.class.getDeclaredMethod("destroy");
+    postConstructDestroy.setAccessible(true);
+    postConstructDestroy.invoke(dummyConsumer);
+  }
+
+  @Test
+  void shouldHandleMessage() {
+    ConsumerRecord<String, String> record = new ConsumerRecord<>("queue.jhipster.dummy", 0, 0, null, "dummy message");
+    boolean actualResult = dummyConsumer.handleMessage(record);
+    assertEquals(true, actualResult);
```
Comment (pascalgrimaud) [inline src/main/resources/generator/server/springboot/broker/kafka/DummyConsumerTest.java.mustache:48]: Use assertj

File: src/main/resources/generator/server/springboot/broker/kafka/DummyConsumerTest.java.mustache [hunk 4]
```diff
@@ -0,0 +1,117 @@
+package {{packageName}}.dummy.infrastructure.primary.kafka.consumer;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
```
Comment (pascalgrimaud) [inline src/main/resources/generator/server/springboot/broker/kafka/DummyConsumerTest.java.mustache:3]: Use assertj instead

File: src/main/resources/generator/server/springboot/broker/kafka/DummyConsumerTest.java.mustache [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    List<ConsumerRecord<String, String>> recordsList = new ArrayList<>();
+    recordsList.add(new ConsumerRecord<>("queue.jhipster.dummy", 0, 0, null, "dummy message"));
+    recordsMap.put(new TopicPartition("queue.jhipster", 0), recordsList);
+    ConsumerRecords<String, String> records = new ConsumerRecords(recordsMap);
+    when(consumer.poll(Duration.ofMillis(10000))).thenReturn(records);
+    dummyConsumer.setClosed(false);
+
+    dummyConsumer.executeKafkaRunner();
+
+    verify(consumer, timeout(3000)).subscribe(Collections.singleton("queue.jhipster.dummy"));
```
Comment (pascalgrimaud) [inline src/main/resources/generator/server/springboot/broker/kafka/DummyConsumerTest.java.mustache:70]: Extract 3000 to local constant

File: src/main/resources/generator/server/springboot/broker/kafka/KafkaConfiguration.java.mustache [hunk 6]
```diff
@@ -8,7 +9,12 @@ import org.springframework.context.annotation.Configuration;
...
 
   @Bean
-  public KafkaProducer<String, String> kafkaProducer(KafkaProducerProperties kafkaProducerProperties) {
-    return new KafkaProducer<>(kafkaProducerProperties.getProducer());
+  public KafkaConsumer<String, String> kafkaConsumer(KafkaProperties kafkaProperties) {
+    return new KafkaConsumer<>(kafkaProperties.getConsumer());
+  }
+
+  @Bean
+  public KafkaProducer<String, String> kafkaProducer(KafkaProperties kafkaProperties) {
```
Comment (pascalgrimaud) [inline src/main/resources/generator/server/springboot/broker/kafka/KafkaConfiguration.java.mustache:17]: It confirms both producer and consumer need to be generated at the same time

File: src/main/resources/generator/server/springboot/broker/kafka/KafkaPropertiesTest.java.mustache [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    ReflectionTestUtils.setField(kafkaProperties, "bootstrapServers", "localhost:9092");
+    ReflectionTestUtils.setField(kafkaProperties, "pollingTimeout", 10_000);
+    Method postConstruct = KafkaProperties.class.getDeclaredMethod("init");
+    postConstruct.setAccessible(true);
+    postConstruct.invoke(kafkaProperties);
+  }
+
+  @Test
+  void shouldGetConsumer() {
+    assertEquals("localhost:9092", kafkaProperties.getConsumer().get(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG));
```
Comment (pascalgrimaud) [inline src/main/resources/generator/server/springboot/broker/kafka/KafkaPropertiesTest.java.mustache:33]: Use assertj instead

File: src/main/resources/generator/server/springboot/broker/kafka/KafkaPropertiesTest.java.mustache [hunk 8]
```diff
@@ -0,0 +1,78 @@
+package {{packageName}}.technical.infrastructure.config.kafka;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
```
Comment (pascalgrimaud) [inline src/main/resources/generator/server/springboot/broker/kafka/KafkaPropertiesTest.java.mustache:3]: Prefer assertj instead

File: tests-ci/generate.sh [hunk 9]
```diff
@@ -232,6 +232,7 @@ elif [[ $application == 'kafkaapp' ]]; then
   callApi "/api/servers/spring-boot/brokers/kafka"
   callApi "/api/servers/spring-boot/brokers/kafka/dummy-producer"
   callApi "/api/servers/spring-boot/brokers/kafka/akhq"
+  callApi "/api/servers/spring-boot/brokers/kafka/dummy-consumer"
```
Comment (pascalgrimaud) [inline tests-ci/generate.sh:235]: Change api to merge producer and consumer

**Review by pascalgrimaud — CHANGES_REQUESTED** [review]

File: tests-ci/generate.sh (hunk 9 above)
Comment (fdelbrayelle) [reply tests-ci/generate.sh:235]: I disagree, you don't want in any case to have the producer AND consumer in the same microservice.

Comment (pascalgrimaud) [reply tests-ci/generate.sh:235]: I'd prefer separate api too, but in this case, you need to rework the test and the configuration class I think

File: src/main/java/tech/jhipster/lite/generator/server/springboot/broker/kafka/domain/KafkaDomainService.java [hunk 10]
```diff
@@ -56,36 +52,47 @@ public void init(final Project project) {
...
-      projectRepository.template(project, SOURCE, "KafkaProducerProperties.java", getPath(MAIN_JAVA, packageNamePath, secondaryKafkaPath));
+      projectRepository.template(project, SOURCE, "DummyProducer.java", getPath(MAIN_JAVA, packageNamePath, DUMMY_PRODUCER_PATH));
+      projectRepository.template(project, SOURCE, "DummyProducerTest.java", getPath(TEST_JAVA, packageNamePath, DUMMY_PRODUCER_PATH));
+      projectRepository.template(project, SOURCE, "DummyProducerIT.java", getPath(TEST_JAVA, packageNamePath, DUMMY_PRODUCER_PATH));
       projectRepository.template(
         project,
         SOURCE,
-        "KafkaProducerPropertiesTest.java",
-        getPath(TEST_JAVA, packageNamePath, secondaryKafkaPath)
+        "KafkaConfiguration.java",
```
Comment (fdelbrayelle) [inline src/main/java/tech/jhipster/lite/generator/server/springboot/broker/kafka/domain/KafkaDomainService.java:83]: Maybe add this in `addConfiguration` above?

Comment (pascalgrimaud) [reply src/main/java/tech/jhipster/lite/generator/server/springboot/broker/kafka/domain/KafkaDomainService.java:83]: Yes it's better

**Review by pascalgrimaud — APPROVED** [review]

**Pull request merged**